### PR TITLE
subclass/widget: Add vfuncs between child_notify and draw

### PR DESCRIPTION
**Pull request by BrainBlasted**
Part of https://github.com/gtk-rs/gtk/issues/862

### CODE REVIEW

**Comment by sdroege** [comment]
Can't review it properly now, will do tomorrow evening.

But gdk::EventAny is basically gdk::Event. And glib::ParamSpecs has bindings, but you'll want to hide the n_pspecs integer. That's the size of the Vec already. You'd use from_glib_none_num() here.

So the two commented out ones should be doable.
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by BrainBlasted** [comment]
How do I translate a `&Vec<glib::ParamSpec>` to `*mut *mut GParamSpec`?

**Comment by sdroege** [comment]
> How do I translate a `&Vec<&glib::ParamSpec>` to `*mut *mut GParamSpec`?

`to_glib_none()` doesn't do that? Alternatively: `vec.iter().map(|p| p.to_glib_none().0).collect::<Vec<_>>()` and then `as_ptr()` on the result.

File: src/subclass/widget.rs [hunk 1]
```diff
@@ -353,6 +716,309 @@ where
...
+    vexpand_ptr: *mut glib_sys::gboolean,
+)
+where
+    T: WidgetImpl,
+{
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Widget = from_glib_borrow(ptr);
+    let mut hexpand_p: bool = *hexpand_ptr != 0;
+    let mut vexpand_p: bool = *vexpand_ptr != 0;
```
Comment (sdroege) [inline src/subclass/widget.rs]: Why not `from_glib()` here?

File: src/subclass/widget.rs [hunk 2]
```diff
@@ -214,6 +311,253 @@ impl<T: WidgetImpl + ObjectImpl> WidgetImplExt for T {
...
+        unsafe {
+            let data = self.get_type_data();
+            let parent_class = data.as_ref().get_parent_class() as *mut gtk_sys::GtkWidgetClass;
+            let f = (*parent_class)
+                .compute_expand
+                .expect("No parent class impl for \"compute_expand\"");
+            f(
+                widget.to_glib_none().0,
+                *hexpand_p as i32 as *mut i32,
+                *vexpand_p as i32 as *mut i32,
```
Comment (sdroege) [inline src/subclass/widget.rs]: I think this is wrong. You're casting an integer to a pointer.

I'd do something like
```rust
let mut h: i32 = hexpand_p.to_glib();
f(&mut h);
*hexpand_p = from_glib(h);
```

File: src/subclass/widget.rs [hunk 3]
```diff
@@ -353,6 +716,309 @@ where
...
+where
+    T: WidgetImpl,
+{
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Widget = from_glib_borrow(ptr);
+    let mut hexpand_p: bool = *hexpand_ptr != 0;
+    let mut vexpand_p: bool = *vexpand_ptr != 0;
+
+    imp.compute_expand(&wrap, &mut hexpand_p, &mut vexpand_p)
```
Comment (sdroege) [inline src/subclass/widget.rs]: Afterwards you need to set the new values of `hexpand_p` / `vexpand_p` to the pointers again

```rust
*hexpand_ptr = hexpand_p.to_glib();
```

**Comment by sdroege** [comment]
Looks good to me otherwise

**Comment by BrainBlasted** [comment]
> > How do I translate a `&Vec<&glib::ParamSpec>` to `*mut *mut GParamSpec`?
> 
> `to_glib_none()` doesn't do that? Alternatively: `vec.iter().map(|p| p.to_glib_none().0).collect::<Vec<_>>()` and then `as_ptr()` on the result.

What's the right way to translate this back?

**Comment by sdroege** [comment]
> What's the right way to translate this back?

You don't have to translate it back. It's an immutable array so it can't be changed anywhere along the call chain.

**Comment by BrainBlasted** [comment]
I need to translate it in order to call from the `widget_dispatch_child_properties_changed` trampoline

**Comment by sdroege** [comment]
An `install_properties()` from `glib/src/subclass/object.rs` for the other way around

**Comment by BrainBlasted** [comment]
So ParamSpec should be owned instead of references?

File: src/subclass/widget.rs (hunk 1 above)
Comment (BrainBlasted) [reply src/subclass/widget.rs]: `error[E0277]: the trait bound `bool: glib::translate::FromGlib<*mut i32>` is not satisfied`

**Comment by BrainBlasted** [comment]
Cleaned up everything according to suggestions.

**Comment by BrainBlasted** [comment]
Rebased and formatted.

File: src/subclass/widget.rs (hunk 1 above)
Comment (sdroege) [reply src/subclass/widget.rs]: `from_glib(*hexpand_ptr)`, you need to dereference

**Comment by sdroege** [comment]
> So ParamSpec should be owned instead of references?

I'd say that it actually should be `&[ParamSpec]` instead of `&Vec<ParamSpec>` as that's more idiomatic Rust.

File: src/subclass/widget.rs (hunk 1 above)
Comment (BrainBlasted) [reply src/subclass/widget.rs]: Yup, figured that out and pushed it.

**Comment by sdroege** [comment]
Yeah looks all good apart from that, thanks :)

**Comment by BrainBlasted** [comment]
Updated to use slices.

**Comment by sdroege** [comment]
@GuillaumeGomez Good to go

**Comment by GuillaumeGomez** [comment]
Thanks!

**Pull request merged**